lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,207 @@
+          #     specifying the attribute name on the model.
+          # @return [Hash]
+          #
+          # @example
+          #   document = {
+          #     data: {
+          #       id: 1,
+          #       type: 'post',
+          #       attributes: {
+          #         title: 'Title 1',
+          #         date: '2015-12-20'
+          #       },
+          #       associations: {
+          #         author: {
+          #           data: {
+          #             type: 'user',
+          #             id: 2
+          #           }
+          #         },
+          #         second_author: {
+          #           data: nil
+          #         },
+          #         comments: {
+          #           data: [{
+          #             type: 'comment',
+          #             id: 3
+          #           },{
+          #             type: 'comment',
+          #             id: 4
+          #           }]
+          #         }
+          #       }
+          #     }
+          #   }
+          #
+          #   parse(document) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   date: '2015-12-20',
+          #     #   author_id: 2,
+          #     #   second_author_id: nil
+          #     #   comment_ids: [3, 4]
+          #     # }
+          #
+          #   parse(document, fields: [:title, :author, date: :published_at]) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   published_at: '2015-12-20',
+          #     #   author_id: 2,
+          #     # }
+          #
+          # rubocop:disable Metrics/CyclomaticComplexity
+          def parse!(document, options = {})
+            fields = parse_fields_option(options[:fields])
+
+            document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)
+
+            fail ArgumentError, 'Invalid payload' unless payload_valid?(document)
+
+            hash = {}
+
+            primary_data = document['data']
+
+            hash[:id] = primary_data['id'] if primary_data['id'] && (fields.nil? || fields[:id])
+
+            attributes = primary_data['attributes']
+            hash.merge!(parse_attributes(attributes, fields)) if attributes
+
+            relationships = primary_data['relationships']
+            hash.merge!(parse_relationships(relationships, fields)) if relationships
+
+            hash
+          end
+          # rubocop:enable Metrics/CyclomaticComplexity
+
+          # Safe version of parse!. Returns an empty hash instead of raising ArgumentError.
+          def parse(document, options = {})
+            parse!(document, options)
+          rescue ArgumentError
+            {}
+          end
+
+          # Checks whether a payload is compliant with the JSON API spec.
+          #
+          # @api private
+          # rubocop:disable Metrics/CyclomaticComplexity
+          def payload_valid?(payload)
```

Comment:
Naming concern: this is the only function in the module that doesn't begin with `parse_`.  and it's usage is `unless payment_valid?` rather than `if payment_invalid?`.

Perhaps this method should be `validate_payload` then return an errors object that is `empty?` when valid.